ipn/ipnlocal/local_test.go

```diff
@@ -3998,3 +4000,535 @@ func TestFillAllowedSuggestions(t *testing.T) {
+	return &notificationWatcher{tb: tb, lb: lb, actor: actor}
+}
+
+func (w *notificationWatcher) watch(mask ipn.NotifyWatchOpt, wanted []wantedNotification, unexpected ...notificationHandler) {
+	w.tb.Helper()
+
+	// Cancel any outstanding [LocalBackend.WatchNotificationsAs] calls.
+	w.mu.Lock()
+	ctxCancel := w.ctxCancel
+	doneCh := w.doneCh
+	w.mu.Unlock()
+	if doneCh != nil {
+		ctxCancel()
+		<-doneCh
+	}
+
+	doneCh = make(chan struct{})
+	gotWantedCh := make(chan struct{})
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	w.tb.Cleanup(func() {
+		ctxCancel()
+		<-doneCh
+	})
+
+	w.mu.Lock()
+	w.mask = mask
+	w.want = wanted
+	w.unexpected = unexpected
+	w.ctxCancel = ctxCancel
+	w.got = nil
+	w.gotWanted = make([]*ipn.Notify, len(wanted))
+	w.gotWantedCh = gotWantedCh
+	w.doneCh = doneCh
+	w.mu.Unlock()
+
+	watchAddedCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		if len(wanted) == 0 {
+			close(gotWantedCh)
+			if len(unexpected) == 0 {
+				close(watchAddedCh)
+				return
+			}
+		}
+
+		var nextWantIdx int
+		w.lb.WatchNotificationsAs(ctx, w.actor, w.mask, func() { close(watchAddedCh) }, func(notify *ipn.Notify) (keepGoing bool) {
+			w.tb.Helper()
+
+			w.lb.logf("receiving...")
+			w.mu.Lock()
+			defer w.mu.Unlock()
+			w.got = append(w.got, notify)
+			w.lb.logf("received(%v): %v", w.actor, notify)
+
+			wanted := false
+			for i := nextWantIdx; i < len(w.want); i++ {
+				if wanted = w.want[i].cond(w.tb, w.actor, notify); wanted {
+					w.gotWanted[i] = notify
+					nextWantIdx = i + 1
+					break
+				}
+			}
+
+			if wanted && nextWantIdx == len(w.want) {
+				close(w.gotWantedCh)
+				if len(w.unexpected) == 0 {
+					// If we have received the last wanted notification,
+					// and we don't have any handlers for the unexpected notifications,
+					// we can stop the watcher right away.
+					return false
+				}
+
+			}
+
+			if !wanted {
+				// If we've received a notification we didn't expect,
+				// it could either be an unwanted notification caused by a bug
+				// or just a miscellaneous one that's irrelevant for the current test.
+				// Call unexpected notification handlers, if any, to
+				// check and fail the test if necessary.
+				for _, h := range w.unexpected {
+					if h(w.tb, w.actor, notify) {
+						break
+					}
+				}
+			}
+
+			return true
+		})
+
+	}()
+	<-watchAddedCh
+}
+
+func (w *notificationWatcher) check() []*ipn.Notify {
+	w.tb.Helper()
+
+	w.mu.Lock()
+	cancel := w.ctxCancel
+	gotWantedCh := w.gotWantedCh
+	checkUnexpected := len(w.unexpected) != 0
+	doneCh := w.doneCh
+	w.mu.Unlock()
+
+	// Wait for up to 10 seconds to receive expected notifications.
+	timeout := 10 * time.Second
+	for {
+		select {
+		case <-gotWantedCh:
+			if checkUnexpected {
+				gotWantedCh = nil
+				// But do not wait longer than a second for unexpected notifications after
+				// the expected notifications have been received.
+				timeout = time.Second
```

Comment:
This makes the tests pretty slow:

```
--- PASS: TestLoginNotifications (4.03s)
    --- PASS: TestLoginNotifications/NoObservers (0.00s)
    --- PASS: TestLoginNotifications/SingleUser (0.00s)
    --- PASS: TestLoginNotifications/SameUser/TwoSessions/NoCID (0.00s)
    --- PASS: TestLoginNotifications/SameUser/TwoSessions/OneWithCID (1.00s)
    --- PASS: TestLoginNotifications/SameUser/TwoSessions/BothWithCID (1.01s)
    --- PASS: TestLoginNotifications/DifferentUsers/NoCID (1.01s)
    --- PASS: TestLoginNotifications/DifferentUsers/SameCID (1.01s)
PASS
ok  	tailscale.com/ipn/ipnlocal	4.045s
```

Instead of waiting a second, can you somehow stop & wait for machinery to settle down?

Alternatively, reduce this timeout to like half a second and t.Parallel all the subtests so they overlap. If this second is not for correctness but just as a window to give it a chance to fail, then that's fine if it'll sometimes miss some failures; with 1 second that's already the case because GitHub Actions is so slow.

Reply:
Good point. Yeah, it's just a window to give an unexpected notification some time to arrive and cause the test to fail.
I changed it to 500ms and enabled parallel execution.